### PR TITLE
PoC for a ProcessorMixin class

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR refactors some common saving/loading functionality of `Processor`s behind a `ProcessorMixin` class which implements the `save_pretrained` and `from_pretrained` method.

It's designed to work with main attributes (by default `feature_extractor` and `tokenizer`, but that list can be changed or extended in subclasses) that are then required by the init and on which `save_pretrained`/`from_pretrained` is called when saving or loading the processor.

It also handles having several classes for a given tokenizer (fast/not fast) so we can have that feature automatically handled. I showed how this works in refactoring two processors (I did not put the tokenizer fast in CLIP because there are some problems with it, but adding it is as easy as changing the `tokenizer_class`).
I will refactor the other processors if the design suits everyone.

### CODE REVIEW

File: docs/source/main_classes/processors.mdx [hunk 1]
```diff
@@ -12,10 +12,22 @@ specific language governing permissions and limitations under the License.
...
+Processors can mean two different things in the Transformers library:
+- the objects that pre-process inputs for multi-modal models such as [Wav2Wec2](../model_doc/wa2vec2) (speech and text)
+  or [CLIP](../model_doc/clip) (text and vision)
+- deprecated objects that were used in older versions of the library to preprocess data for GLUE or SQUAD.
 
-## Processors
+## Multi-modal processors
+
+Any multi-modal model will require an object to encode or decode the data that groups several modalities (among text,
+vision and speech). This is handled by objects called processors, which group tokenizer (for the text modaility) and
```
Comment (NielsRogge) [inline docs/source/main_classes/processors.mdx:23]:
```suggestion
Any multi-modal model will require an object to encode or decode the data that groups several modalities (among text,
vision and audio). This is handled by objects called processors, which group tokenizer (for the text modality) and
```
For me, speech is only one type of audio.

File: src/transformers/processing_utils.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        if len(args) != len(self.attributes):
+            raise ValueError(
+                f"This processor requires {len(self.attributes)} arguments: {', '.join(self.attributes)}. Got "
+                f"{len(args)} arguments instead."
+            )
+
+        # Check each arg is of the proper class (this will also catch a user initializing in the wrong order)
+        for arg, attribute_name in zip(args, self.attributes):
+            class_name = getattr(self, f"{attribute_name}_class")
+            # Nothing is every going to be an instance of "AutoXxx", in that case we check the base class.
```
Comment (patil-suraj) [inline src/transformers/processing_utils.py:57]:
```suggestion
            # Nothing is ever going to be an instance of "AutoXxx", in that case we check the base class.
```

**Review by patil-suraj — APPROVED** [review]
Very cool! This looks good to me.

File: docs/source/main_classes/processors.mdx [hunk 3]
```diff
@@ -12,10 +12,22 @@ specific language governing permissions and limitations under the License.
 
 # Processors
 
-This library includes processors for several traditional tasks. These processors can be used to process a dataset into
-examples that can be fed to a model.
+Processors can mean two different things in the Transformers library:
```
Comment (patrickvonplaten) [inline docs/source/main_classes/processors.mdx:15]: Nice!

Comment (LysandreJik) [reply docs/source/main_classes/processors.mdx:15]: Thanks for reworking this part of the docs!

**Review by LysandreJik — APPROVED** [review]
Looks good to me!

File: src/transformers/processing_utils.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+            class_name = AUTO_TO_BASE_CLASS_MAPPING.get(class_name, class_name)
+            if isinstance(class_name, tuple):
+                proper_class = tuple(getattr(transformers_module, n) for n in class_name if n is not None)
+            else:
+                proper_class = getattr(transformers_module, class_name)
+
+            if not isinstance(arg, proper_class):
+                raise ValueError(
```
Comment (patrickvonplaten) [inline src/transformers/processing_utils.py:65]: Great error message

File: src/transformers/processing_utils.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+            save_directory (`str` or `os.PathLike`):
+                Directory where the feature extractor JSON file and the tokenizer files will be saved (directory will
+                be created if it does not exist).
+        """
+        for attribute_name in self.attributes:
+            attribute = getattr(self, attribute_name)
+            # Include the processor class in the attribute config so this processor can then be reloaded with the
+            # `AutoProcessor` API.
+            if hasattr(attribute, "_set_processor_class"):
+                attribute._set_processor_class(self.__class__.__name__)
```
Comment (patrickvonplaten) [inline src/transformers/processing_utils.py:99]: Should we add a test to make sure that every tokenizer & feature extractor has this function in a follow up PR?

File: docs/source/main_classes/processors.mdx [hunk 6]
```diff
@@ -12,10 +12,22 @@ specific language governing permissions and limitations under the License.
...
+- the objects that pre-process inputs for multi-modal models such as [Wav2Wec2](../model_doc/wa2vec2) (speech and text)
```
Comment (NielsRogge) [inline docs/source/main_classes/processors.mdx:16]:
```suggestion
- the objects that pre-process inputs for multi-modal models such as [Wav2Vec2](../model_doc/wav2vec2) (speech and text)
```

File: docs/source/main_classes/processors.mdx (hunk 1 above)
Comment (NielsRogge) [inline docs/source/main_classes/processors.mdx:23]:
```suggestion
vision and speech). This is handled by objects called processors, which group tokenizers (for the text modaility) and
```

File: docs/source/main_classes/processors.mdx [hunk 7]
```diff
@@ -12,10 +12,22 @@ specific language governing permissions and limitations under the License.
...
+  or [CLIP](../model_doc/clip) (text and vision)
+- deprecated objects that were used in older versions of the library to preprocess data for GLUE or SQUAD.
 
-## Processors
+## Multi-modal processors
+
+Any multi-modal model will require an object to encode or decode the data that groups several modalities (among text,
+vision and speech). This is handled by objects called processors, which group tokenizer (for the text modaility) and
+feature extractors (for vision and speech).
```
Comment (NielsRogge) [inline docs/source/main_classes/processors.mdx:24]:
```suggestion
feature extractors (for vision and audio).
```

File: src/transformers/processing_utils.py (hunk 5 above)
Comment (sgugger) [reply src/transformers/processing_utils.py:99]: It is defined by the base classes (`FeatureExtractorMixin` and `PreTrainedTokenizerBase`) so I don't think it's necessary.

File: src/transformers/processing_utils.py [hunk 8]
```diff
@@ -44,15 +44,25 @@ class ProcessorMixin:
...
+        for key in kwargs:
+            if key not in self.attributes:
+                raise TypeError(f"Unexepcted keyword argument {key}.")
+        for arg, attribute_name in zip(args, self.attributes):
+            if attribute_name in kwargs:
+                raise TypeError(f"Got multiple values for argument {attribute_name}.")
+            else:
+                kwargs[attribute_name] = arg
+
+        if len(kwargs) != len(self.attributes):
```
Comment (LysandreJik) [inline src/transformers/processing_utils.py:58]: That's smart!

**Review by LysandreJik — APPROVED** [review]

File: src/transformers/processing_utils.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+# limitations under the License.
+"""
+ Processing saving/loading class for common processors.
+"""
+
+import importlib.util
+from pathlib import Path
+
+
+# Comment to write
```
Comment (kayhayen) [inline src/transformers/processing_utils.py:23]: Why did you find the need to duplicate the `transformers` module in memory. This code executes it again, and instists on how it is loaded, for no obvious reason.

I do not recognize what the difference is to `import transformers as transformers_module` can you explain?

**Pull request merged**